### DISCUSSION TITLE
Mirroring images

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#642] [Q&A]
Hi @pgundlach, 

Happy new year! 🍾😁

I wonder if, for the next year, you could add an attribute _mirror_ for the image tag, e.g.: `mirror="vertical/horizontal"` or `mirror="x-axis/y-axis"` ; meanwhile I am trying with the transformation matrix and, as last resort, just flipping the images... 😓

See you next year! 😇

**Comment by pgundlach** [comment]
Horizontal mirroring:

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <Record element="data">
        <PlaceObject>
            <Image width="3cm" file="ocean.pdf" />
        </PlaceObject>
        <PlaceObject>
            <Transformation matrix="-1 0 0 1 0 0">
                <Image width="3cm" file="ocean.pdf" />
            </Transformation>
        </PlaceObject>
    </Record>
</Layout>
```

Vertical mirroring (not quite what I expected)

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <Record element="data">
        <PlaceObject>
            <Image width="3cm" file="ocean.pdf" />
        </PlaceObject>
        <PlaceObject frame="solid" framecolor="green">
            <Transformation matrix="1 0 0 -1 0 0">
                <Image width="3cm" file="ocean.pdf" />
            </Transformation>
        </PlaceObject>
    </Record>
</Layout>
```

I need to look into what happens with vertical mirroring

**Reply by Cicorione** [reply]
Regarding this:

`<Transformation matrix="-1 0 0 1 0 0">`

This is the first thing I tried, but the background wasn't visible on the page, the background measure the full page, thus I think it was wrong...

But now that you showed me this I do believe that it mirrored from coordinate `x=0` and not from its center, so perhaps the background landed outside the page, that is why I couldn't see it... 🤔

**Comment by pgundlach** [comment]
Happy new year BTW!

**Comment by pgundlach** [comment]
To add: this is not my solution, just a thought. I am still thinking about a good interface. 

I am currently thinking about something like this: 

```xml
<Transformation mirror-h="yes" mirror-v="yes">
        <Image width="3cm" file="ocean.pdf" />
</Transformation>
```

**Reply by Cicorione** [reply]
I am not wise like you, but I do prefer succinct attribute; what about with an attribute with two values:

* `mirror="yes,yes"`
* `mirror="1,1"`

Also, wouldn't it be better as attribute of `Image` rather than `Transformation`? `Image` already has `rotate` as attribute, rotate and mirror are complimentary. 

Thanks!

**Reply by pgundlach** [reply]
You are right with the succinct attributes. Perhaps something like "horizontal", "vertical" and "both" could be valid options? And naming the attribute: should it be 'mirror' or 'flip'? Or do you have a better suggestion?

The advantage of having the mirroring with `<Transformation>` is that you a mirror any object, not only images. The rotate attribute is somewhat special as it uses PDF operators that apply to images only. 

It would not harm to have a mirroring option on both (`<Image>`  and `<Transformation>`)

I still need to check if clipping etc. needs special considerations.

**Reply by Cicorione** [reply]
I did a quick research and both Illustrator and Inkscape call it **FLIP**, so I guess this is the common used name.

I agree with you that having all the "transformations" in one place is more predictable and consistent than having in two places, this avoids to have situation where you are applying two transformations even though your image looks correctly oriented.

**Comment by pgundlach** [comment]
Just a note for me... 

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">
    <Pageformat height="20cm" width="7cm"></Pageformat>

    <Record element="data">
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value>Original:</Value>
                </Paragraph>
            </Textblock>
        </PlaceObject>
        <PlaceObject>
            <Image width="3cm" file="ocean.pdf" />
        </PlaceObject>
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value>Mirror vertical:</Value>
                </Paragraph>
            </Textblock>
        </PlaceObject>
        <PlaceObject frame="solid" framecolor="green">
            <Transformation mirror-y="yes">
                <Image width="3cm" file="ocean.pdf" />
            </Transformation>
        </PlaceObject>
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value>Mirror horizontal:</Value>
                </Paragraph>
            </Textblock>
        </PlaceObject>
        <PlaceObject frame="solid" framecolor="green">
            <Transformation mirror-x="yes">
                <Image width="3cm" file="ocean.pdf" />
            </Transformation>
        </PlaceObject>
        <NextRow />
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value>Mirror both axes:</Value>
                </Paragraph>
            </Textblock>
        </PlaceObject>
        <PlaceObject frame="solid" framecolor="green">
            <Transformation mirror-x="yes" mirror-y="yes">
                <Image width="3cm" file="ocean.pdf" />
            </Transformation>
        </PlaceObject>
    </Record>
</Layout>
```

**Comment by pgundlach** [comment]
Now implemented in 4.19.37.

**Reply by Cicorione** [reply]
Great!

Thank you very much!!!

**Comment by pgundlach** [comment]
https://doc.speedata.de/publisher/en/basics/outputtingobjects/#ch-outputtingobjectds-flipping
